synapse/replication/resource.py

```diff
@@ -274,11 +274,18 @@ def presence(self, writer, current_token, request_streams):
 
     @defer.inlineCallbacks
     def typing(self, writer, current_token, request_streams):
-        current_position = current_token.presence
+        current_position = current_token.typing
 
         request_typing = request_streams.get("typing")
 
         if request_typing is not None:
+            # If they have a higher token than current max, we can assume that
+            # they had been talking to a previous instance of the master. Since
+            # we reset the token on restart, the best (but hacky) thing we can
+            # do is to simply resend down all the typing notifications.
+            if request_typing > current_position:
+                request_typing = 0
```

Comment:
In particular, there is nowhere in the slaves where do a `stream_token = max(last_token, new_token)`, we always just use what was given to us in the response.